_drafts/practical_zephyr_01_kconfig.md

```diff
@@ -0,0 +1,986 @@
+
+```bash
+$ rm -rf ../build
+$ west build --board nrf52840dk_nrf52840 -d ../build -- \
+  -DCONF_FILE="prj_release.conf" \
+  -DEXTRA_CONF_FILE="extra1.conf;extra0.conf"
+```
+```
+Parsing /opt/nordic/ncs/v2.4.0/zephyr/Kconfig
+Loaded configuration '/opt/nordic/ncs/v2.4.0/zephyr/boards/arm/nrf52840dk_nrf52840/nrf52840dk_nrf52840_defconfig'
+Merged configuration '/path/to/01_kconfig/prj_release.conf'
+Merged configuration '/path/to/01_kconfig/boards/nrf52840dk_nrf52840_release.conf'
+Merged configuration '/path/to/01_kconfig/extra1.conf'
+Merged configuration '/path/to/01_kconfig/extra0.conf'
+Configuration saved to '/path/to/build/zephyr/.config'
+```
+
+
+
+## Kconfig hardening
+
+Another tool worth mentioning when talking about _Kconfig_ is the [hardening tool](https://docs.zephyrproject.org/latest/security/hardening-tool.html). Just like `menuconfig` and `guiconfig`, this tool is a target for `west build`. It is executed using the `-t hardenconfig` target in the call to `west build`. The hardening tool checks the _Kconfig_ symbols against a set of known configuration options that should be used for a secure Zephyr application and lists all differences found in the application.
+
+We can use this tool to check our normal and _release_ configurations:
+
+```bash
+$ # test normal build
+$ west build --board nrf52840dk_nrf52840 \
+  -d ../build \
+  --pristine \
+  -t hardenconfig
+
+$ # test release build
+$ west build --board nrf52840dk_nrf52840 \
+  -d ../build \
+  --pristine \
+  -t hardenconfig \
+  -- -DCONF_FILE=prj_release.conf
+```
+
+For the normal build using `prj.conf`, the hardening tool displays a table of all symbols whose current values do not match the recommended, secure configuration value. E.g., at the time of writing this is the output for the normal build in my demo application:
+
+```
+-- west build: running target hardenconfig
+[0/1] cd /path/to/zephyr/kconfig/hardenconfig.py /opt/nordic/ncs/v2.4.0/zephyr/Kconfig
+                 name                 | current | recommended || check result
+==============================================================================
+CONFIG_OVERRIDE_FRAME_POINTER_DEFAULT |    n    |      y      ||     FAIL
+CONFIG_USE_SEGGER_RTT                 |    y    |      n      ||     FAIL
+CONFIG_BUILD_OUTPUT_STRIPPED          |    n    |      y      ||     FAIL
+CONFIG_FAULT_DUMP                     |    2    |      0      ||     FAIL
+CONFIG_STACK_SENTINEL                 |    n    |      y      ||     FAIL
+```
+
+The symbols in `prj_release.conf` have been chosen such that at the time of writing all hardening options are fulfilled and the above table is empty.
+
+
+
+## A custom Kconfig symbol
+
+In the previous sections, we had a thorough look at _Kconfig_ and its workings. To wrap up this article, we'll create a custom, application-specific symbol and use it in our application _and_ build process. While this is a rather unlikely use case, it nicely demonstrates how _Kconfig_ works.
```

Comment:
I would remove the last sentence here. Speaking from experience, this is another feature of Zephyr that I think is really great. It has helped prior teams quickly enable different test features or build different configs specifically related to our device.

Reply:
Good point, I'll remove it.